File: modules/action/src/main/scala/action/async/AsyncSpec.scala
package action.async

import scala.concurrent.Future

final case class AsyncOutput[I, K, O, R](outputDecoder: O => Option[Either[Throwable, R]],
                                         serdes: AsyncSerdes[K, R],
                                         topicName: I => Option[String],
                                         topicNames: List[String])

/**
  * @param actionType
  * @param inputDecoder
  * @param keyMapper
  * @param asyncFunction
  * @param groupId
  * @param outputSpec
  * @tparam A - common representation form for all action commands (typically Json / GenericRecord for Avro)
  * @tparam I - input to async function
  * @tparam K - key for the output topic
  * @tparam O - output returned by async function
  * @tparam R - final result type that ends up in output topic
  */
final case class AsyncSpec[A, I, K, O, R](actionType: String,
                                          inputDecoder: A => Either[Throwable, I],
                                          keyMapper: I => K,
                                          asyncFunction: I => Future[O],
                                          groupId: String,
                                          outputSpec: Option[AsyncOutput[I, K, O, R]])

File: modules/action/src/main/java/io/simplesource/saga/action/async/AsyncSpec.java
package io.simplesource.saga.action.async;

import io.simplesource.data.Result;
import lombok.Value;

import java.util.Optional;
import java.util.function.BiConsumer;
import java.util.function.Function;

/**
  * @param <A> - common representation form for all action commands (typically Json / GenericRecord for Avro)
  * @param <I> - input to async function
  * @param <K> - key for the output topic
  * @param <O> - output returned by async function
  * @param <R> - final result type that ends up in output topic
  */
@Value
final public class AsyncSpec<A, I, K, O, R> {
    public final String actionType;
    public final Function<A, Result<Throwable, I>> inputDecoder;
    public final Function<I, K> keyMapper;
    public final BiConsumer<I, CallBack<O>> asyncFunction;
    public final String groupId;
    public final Optional<AsyncOutput<I, K, O, R>> outputSpec;
}
